pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+		// zero weight at the beginning or at the end of a sale may cause a problem in the price calculation
+		ensure!(
+			!pool_data.initial_weights.0.is_zero()
+				&& !pool_data.initial_weights.1.is_zero()
+				&& !pool_data.final_weights.0.is_zero()
+				&& !pool_data.final_weights.1.is_zero(),
+			Error::<T>::ZeroWeight
+		);
+
+		Ok(())
+	}
+
+	/// return true if now is in interval <pool.start, pool.end> WARN: pool.paused DOESN'T MATTER
+	fn is_pool_running(pool_data: &Pool<T::AccountId, T::BlockNumber>) -> bool {
+		let now = <frame_system::Pallet<T>>::block_number();
+		pool_data.start <= now && now <= pool_data.end
+	}
+
+	/// return true if now is in interval <pool.start, pool.end> and POOL IS NOT PAUSED
+	fn is_sale_running(pool_data: &Pool<T::AccountId, T::BlockNumber>) -> bool {
+		Self::is_pool_running(pool_data) && !pool_data.paused
+	}
+	/// return true if the LBP event has not yet started, or the beginning is not set
+	fn is_prior_sale_or_uninitialized(pool_data: &Pool<T::AccountId, T::BlockNumber>) -> bool {
+		let now = <frame_system::Pallet<T>>::block_number();
+		pool_data.start.is_zero() || now < pool_data.start
+	}
+
+	fn is_after_sale(pool_data: &Pool<T::AccountId, T::BlockNumber>) -> bool {
+		let now = <frame_system::Pallet<T>>::block_number();
+		pool_data.end >= now
+	}
+
+	fn update_weights_and_validate_trade(
+		who: &T::AccountId,
+		assets: AssetPair,
+		amount: BalanceOf<T>,
+		limit: BalanceOf<T>,
+		trade_type: TradeType,
+	) -> Result<AMMTransfer<T::AccountId, AssetId, AssetPair, Balance>, DispatchError> {
+		ensure!(!amount.is_zero(), Error::<T>::ZeroAmount);
+
+		ensure!(
+			T::MultiCurrency::free_balance(assets.asset_in, &who) >= amount,
+			Error::<T>::InsufficientAssetBalance
+		);
+
+		let pool_id = Self::get_pair_id(assets);
+		let mut pool_data = <PoolData<T>>::try_get(&pool_id).map_err(|_| Error::<T>::PoolNotFound)?;
+
+		ensure!(Self::is_sale_running(&pool_data), Error::<T>::SaleIsNotRunning);
+
+		// update weights or reuse the last if update is not necessary
+		let (weight_in, weight_out) = match Self::update_weights(&pool_id, &mut pool_data) {
+			Ok(weights) => {
+				if assets.asset_in == pool_data.assets.0 {
+					(weights.0, weights.1)
+				} else {
+					// swap weights if assets are in different order
+					(weights.1, weights.0)
+				}
+			}
+			Err(_) => return Err(<Error<T>>::Overflow.into()),
+		};
+
+		let asset_in_reserve = T::MultiCurrency::free_balance(assets.asset_in, &pool_id);
+		let asset_out_reserve = T::MultiCurrency::free_balance(assets.asset_out, &pool_id);
+		if trade_type == TradeType::Sell {
+			ensure!(
+				amount <= asset_in_reserve / MAX_IN_RATIO,
+				Error::<T>::MaxInRatioExceeded
+			);
+		} else {
+			ensure!(
+				amount <= asset_out_reserve / MAX_OUT_RATIO,
+				Error::<T>::MaxOutRatioExceeded
+			);
+		}
+
+		let (amount_in, amount_out, fee_asset, fee_amount) = if trade_type == TradeType::Sell {
+			let token_amount_out = hydra_dx_math::lbp::calculate_out_given_in(
```

Comment:
fee should be part of this calculation, so the frontend can return the correct value, you can return transfer fee as second return value for transfer

Reply:
As discussed, we are not going to add transfer fee to hydra_dx_math package because it brings new dependencies.